### PR TITLE
PeeringAcceptor controller

### PR DESCRIPTION
**Pull request by ndhanushkodi**
Changes proposed in this PR:
- PeeringAcceptor CR + controller + unit tests
- was NOT planning to unit test getExistingSecret or createK8sPeeringTokenSecretWithOwner because their cases are covered in the ReconcileCreateUpdate test. LMK if you see gaps or prefer the helper function tests.
- was NOT planning to unit test the generateToken, deletePeering, or createSecret helper functions as they are heavily used throughout other tests.

Future PRs (tasks are in Cluster Peering Asana):
- PeeringDialer CR + controller
- upstream changes for CRDs and upstream annotation
- Acceptance test for service mesh
- Version annotation (and storing that as peering metadata)
- Adding Finalizer
- Add peering state/status to the CR status

How I've tested this PR:
`helm install nitya ../../consul-k8s/charts/consul -f consul-values.yml`

`consul-values.yml`:
```yaml
global:
  domain: consul
  datacenter: dc1
  imageK8S: ndhanushkodi/consul-k8s-dev:peering22
  image: ndhanushkodi/consul-dev:peering4
server:
  replicas: 1
  bootstrapExpect: 1
client:
  enabled: true
  grpc: true
ui:
  enabled: true
connectInject:
  enabled: true
controller:
  enabled: true

```

`k apply -f peering-acceptor.yaml`

```
apiVersion: consul.hashicorp.com/v1alpha1
kind: PeeringAcceptor
metadata:
  name: peeraaaaa
spec:
  peer:
    secret:
      name: "foo"
      key: "bar"
      backend: "kubernetes"
```

`k delete -f peering-acceptor.yaml`

Look at the logs and make sure it's as expected.

How I expect reviewers to test this PR:
Code review

Checklist:
- [x] Tests added
- [-] CHANGELOG entry added 
  > HashiCorp engineers only, community PRs should not add a changelog entry.
  > Entries should use present tense (e.g. Add support for...)

### CODE REVIEW

File: control-plane/connect-inject/peering_acceptor_controller.go [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+		}
+
+	}
+	return false, nil
+}
+func (r *PeeringAcceptorController) updateStatus(ctx context.Context, peeringAcceptor *consulv1alpha1.PeeringAcceptor, resp *api.PeeringGenerateTokenResponse) error {
+	peeringAcceptor.Status.Secret = &consulv1alpha1.SecretStatus{
+		Name:    peeringAcceptor.Spec.Peer.Secret.Name,
+		Key:     peeringAcceptor.Spec.Peer.Secret.Key,
+		Backend: peeringAcceptor.Spec.Peer.Secret.Backend,
```
Comment (thisisnotashwin) [inline control-plane/connect-inject/peering_acceptor_controller.go:164]: some of these can be simplified by adding methods to the PeeringAcceptor like SecretName() that will return spec.peer.secret.name so that it is more readable and you get the benefits of adding Nil checks.

**Review by thisisnotashwin — COMMENTED** [review]
I like what im seeing so far

File: control-plane/api/v1alpha1/peeringacceptor_types.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+	Secret *Secret `json:"secret,omitempty"`
+	//Generator *Generator `json:"generator,omitempty"`
+	//Requester *Requester `json:"requester,omitempty"`
+}
+type Generator struct {
+	Secret *Secret `json:"secret,omitempty"`
+}
+type Requester struct {
+	Secret *Secret `json:"secret,omitempty"`
+}
```
[the remark this answers is not in the document]
Comment (ndhanushkodi) [reply control-plane/api/v1alpha1/peeringacceptor_types.go:54]: Yep! Thanks, just cleaned them up

File: control-plane/connect-inject/peering_acceptor_controller.go [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+		r.Log.Error(err, "failed to get Peering from Consul", "name", req.Name)
+		return ctrl.Result{}, err
+	}
+
+	// TODO(peering): Verify that the existing peering in Consul is an acceptor peer. If it is a dialing peer, an error should be thrown.
+	// If the peering does exist in Consul, compare the existing status to the spec, and decide whether to make updates.
+
+	// TODO(peering): If the status is empty, then what should be done? This means a peering exists in Consul but hasn't
+	// been reconciled by the controller, or there was an error during a previous reconciliation. I think this means we
+	// should continue and generate a new token and update the status with that.
```
[the remark this answers is not in the document]
Comment (ndhanushkodi) [reply control-plane/connect-inject/peering_acceptor_controller.go:104]: I like this idea and have added a ReconcileError to the status! In this case though, we want to know whether we have enough information in the status to get the previous secret value. I think even if we don't have a secret set in the status, we do want to generate a new token, because it means we've had an error in a previous reconcile or that the status update itself has failed. So I think if we do reach this case, we just do want to generate a new token. I've updated this behavior to do that

File: control-plane/connect-inject/peering_acceptor_controller.go [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+	if peeringAcceptor.Status.Secret == nil || peeringAcceptor.Status.LastReconcileTime == nil {
+		return false, errors.New("shouldGenerateToken was called with an empty fields in the existing status")
+	}
+	// Compare the existing name, key, and backend.
+	if peeringAcceptor.Status.Secret.Name != peeringAcceptor.Spec.Peer.Secret.Name {
+		return true, nil
+	}
+	if peeringAcceptor.Status.Secret.Key != peeringAcceptor.Spec.Peer.Secret.Key {
+		return true, nil
+	}
```
[the remark this answers is not in the document]
Comment (ndhanushkodi) [reply control-plane/connect-inject/peering_acceptor_controller.go:142]: I saw it happen automatically-- and it's tested in a line that looks like this in the tests:
```
require.Equal(t, "", createdSecret.StringData["some-old-key"])
```
There's 2 test cases for existing secrets-- one when the secret there without being created by the controller and the other when the secret is managed by the controller.

File: control-plane/connect-inject/peering_acceptor_controller.go [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+}
+
+func (r *PeeringAcceptorController) shouldGenerateToken(peeringAcceptor *consulv1alpha1.PeeringAcceptor, existingSecret *corev1.Secret) (bool, error) {
+	if peeringAcceptor.Status.Secret == nil || peeringAcceptor.Status.LastReconcileTime == nil {
+		return false, errors.New("shouldGenerateToken was called with an empty fields in the existing status")
+	}
+	// Compare the existing name, key, and backend.
+	if peeringAcceptor.Status.Secret.Name != peeringAcceptor.Spec.Peer.Secret.Name {
+		return true, nil
+	}
```
[the remark this answers is not in the document]
Comment (ndhanushkodi) [reply control-plane/connect-inject/peering_acceptor_controller.go:139]: Yup! Handled this case now

File: control-plane/connect-inject/peering_acceptor_controller.go (hunk 1 above)
Comment (ndhanushkodi) [reply control-plane/connect-inject/peering_acceptor_controller.go:165]: This is a great idea, going to do a passthrough to do this.

File: charts/consul/templates/connect-inject-clusterrole.yaml [hunk 6]
```diff
@@ -24,6 +24,13 @@ rules:
   - "get"
   - "list"
   - "watch"
+- apiGroups: [ "" ]
+  resources: ["secrets"]
+  verbs:
+    - "get"
+    - "list"
+    - "watch"
+    - "create"
```
[the remark this answers is not in the document]
Comment (ndhanushkodi) [reply charts/consul/templates/connect-inject-clusterrole.yaml:33]: Ahh yes! The unit tests didn't catch that but I think once I manually test the secret deletion case it will! 

For the ownerRef garbage collection, we don't need delete access, but for the case where the secret name was changed we will!

File: control-plane/api/v1alpha1/peeringacceptor_types.go [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+	LastReconcileTime *metav1.Time  `json:"lastReconcileTime,omitempty" description:"last time the resource was reconciled"`
+	Secret            *SecretStatus `json:"secret,omitempty"`
+}
+type SecretStatus struct {
+	// TODO(peering): add additional status fields
+	Name       string                  `json:"name,omitempty"`
+	Key        string                  `json:"key,omitempty"`
+	Backend    string                  `json:"backend,omitempty"`
+	LatestHash string                  `json:"latestHash,omitempty"`
+	Kubernetes *KubernetesSecretStatus `json:"kubernetes,omitempty"`
```
[the remark this answers is not in the document]
Comment (ndhanushkodi) [reply control-plane/api/v1alpha1/peeringacceptor_types.go:77]: hey sorry wasn't sure what you meant here

File: control-plane/connect-inject/peering_acceptor_controller.go [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+
+	// If the PeeringAcceptor resource has been deleted (and we get an IsNotFound
+	// error), we need to delete it in Consul.
+	if k8serrors.IsNotFound(err) {
+		r.Log.Info("PeeringAcceptor was deleted, deleting from Consul", "name", req.Name, "ns", req.Namespace)
+		_, err := r.deletePeering(ctx, req.Name)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
```
[the remark this answers is not in the document]
Comment (ndhanushkodi) [reply control-plane/connect-inject/peering_acceptor_controller.go:64]: So my understanding from the finalizer is that it will cause an event so that I could have access to the whole PeeringAcceptor before it is truly deleted, so that I can make sure it is deleted in Consul. I think this case is handled by the existing logic though, right? Since we only need the event's req.Name, even if deletion in Consul fails, it'll get requeued with the same name, and attempt to retry the delete in Consul which is what we want? LMK if I'm missing something!

File: control-plane/connect-inject/peering_acceptor_controller.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (thisisnotashwin) [reply control-plane/connect-inject/peering_acceptor_controller.go]: that makes a lot of sense!

File: control-plane/connect-inject/peering_acceptor_controller.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (thisisnotashwin) [reply control-plane/connect-inject/peering_acceptor_controller.go]: brilliant!

File: control-plane/api/v1alpha1/peeringacceptor_types.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (ndhanushkodi) [reply control-plane/api/v1alpha1/peeringacceptor_types.go:77]: Ahh got it! Also so far I'm not using this field, since the Name Key and Backend are all I need to get the secret and compare it to the spec, do you think its better practice to use the object reference instead of the secret.name field?

File: .github/workflows/test.yml [hunk 9]
```diff
@@ -142,14 +142,22 @@ jobs:
       - run:  mkdir -p ${{env.TEST_RESULTS}}
```
Comment (ndhanushkodi) [inline .github/workflows/test.yml:142]: The changes in this file would be removed when merging in the feature branch. This is because the most recently released version of consul does not have peering.

File: control-plane/connect-inject/peering_acceptor_controller.go [hunk 10]
```diff
@@ -0,0 +1,335 @@
...
+//+kubebuilder:rbac:groups=apps,resources=secrets,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=apps,resources=secrets/status,verbs=get
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// PeeringAcceptor resources determine whether to generate a new peering token in Consul and store it in the backend
+// specified in the spec. If the resource doesn't exist, the peering should be deleted in Consul. If the resource
+// exists, and a peering doesn't exist in Consul, it should be created. If the resource exists, and a peering does exist
+// in Consul, it should be reconciled. If the status of the resource does not match the current state of the specified
+// secret, generate a new token and store it according to the spec.
```
Comment (t-eckert) [inline control-plane/connect-inject/peering_acceptor_controller.go:45]: Might be easier to parse as a list? 🤷🏻 

```suggestion
// PeeringAcceptor resources determine whether to generate a new peering token in Consul and store it in the backend
// specified in the spec. 
//  - If the resource doesn't exist, the peering should be deleted in Consul. 
//  - If the resource exists and a peering doesn't exist in Consul, it should be created. 
//  - If the resource exists and a peering does exist in Consul, it should be reconciled. 
//  - If the status of the resource does not match the current state of the specified
//    secret, generate a new token and store it according to the spec.
```

File: control-plane/connect-inject/peering_acceptor_controller.go [hunk 11]
```diff
@@ -0,0 +1,335 @@
...
+		return ctrl.Result{}, nil
+	} else if err != nil {
+		r.Log.Error(err, "failed to get PeeringAcceptor", "name", req.Name, "ns", req.Namespace)
+		return ctrl.Result{}, err
+	}
+
+	statusSecretSet := false
+	if peeringAcceptor.Status.Secret != nil {
+		statusSecretSet = true
+	}
```
Comment (t-eckert) [inline control-plane/connect-inject/peering_acceptor_controller.go:73]: This could be inlined and made more clearly a bool with 
```suggestion
	isStatusSecretSet := (peeringAcceptor.Status.Secret != nil)
```

**Comment by t-eckert** [comment]
I don't feel I have much to add given my lack of context. I perused to really just gain a better understanding of what you are working on here. I think you've expressed it very clearly.

**Review by ishustava — COMMENTED** [review]
Great work, Nitya! As usual, love all the comments and the thought you put into this ❤️ 

I'm leaving my comments so far (I got to reviewing most of the reconcile logic, except for some private functions and tests).

I will review some more today, but wanted to leave my comments so far. Some of them are very minor and I marked them as non-blockers. Others are more questions or thoughts about edge cases, so feel free to punt on those too if need be!

File: control-plane/connect-inject/peering_acceptor_controller.go [hunk 12]
```diff
@@ -0,0 +1,335 @@
...
+		Key:     peeringAcceptor.Spec.Peer.Secret.Key,
+		Backend: peeringAcceptor.Spec.Peer.Secret.Backend,
+	}
+
+	peeringTokenHash := sha256.Sum256([]byte(resp.PeeringToken))
+	peeringAcceptor.Status.Secret.LatestHash = hex.EncodeToString(peeringTokenHash[:])
+
+	peeringAcceptor.Status.LastReconcileTime = &metav1.Time{Time: time.Now()}
+	peeringAcceptor.Status.ReconcileError = &consulv1alpha1.ReconcileErrorStatus{
+		Error:   pointerToBool2(false),
```
Comment (ishustava) [inline control-plane/connect-inject/peering_acceptor_controller.go:214]: should we use `pointerToBool` (I'm assuming we already have this function in this package - that's why it's named like this?)

File: control-plane/connect-inject/peering_acceptor_controller.go [hunk 13]
```diff
@@ -0,0 +1,335 @@
...
+		existingSecretHash := hex.EncodeToString(existingSecretHashBytes[:])
+		if existingSecretHash != peeringAcceptor.Status.Secret.LatestHash {
+			return true, false, nil
+		}
+
+	}
+	return false, false, nil
+}
+
+// updateStatusError updates the peeringAcceptor's secret in the status.
```
Comment (ishustava) [inline control-plane/connect-inject/peering_acceptor_controller.go:201]:
```suggestion
// updateStatus updates the peeringAcceptor's secret in the status.
```

File: control-plane/connect-inject/peering_acceptor_controller.go [hunk 14]
```diff
@@ -0,0 +1,335 @@
...
+	}
+	err := r.Status().Update(ctx, peeringAcceptor)
+	if err != nil {
+		r.Log.Error(err, "failed to update PeeringAcceptor status", "name", peeringAcceptor.Name, "namespace", peeringAcceptor.Namespace)
+	}
+	return err
+}
+
+// updateStatusError updates the peeringAcceptor's ReconcileError in the status.
+func (r *PeeringAcceptorController) updateStatusError(ctx context.Context, peeringAcceptor *consulv1alpha1.PeeringAcceptor, reconcileErr error) error {
```
Comment (ishustava) [inline control-plane/connect-inject/peering_acceptor_controller.go:225]: We always ignore the error that's returned by this function. I'm assuming we shouldn't, so maybe it could be a TODO?

File: control-plane/connect-inject/peering_acceptor_controller.go [hunk 15]
```diff
@@ -0,0 +1,335 @@
...
+	err := r.Status().Update(ctx, peeringAcceptor)
+	if err != nil {
+		r.Log.Error(err, "failed to update PeeringAcceptor status", "name", peeringAcceptor.Name, "namespace", peeringAcceptor.Namespace)
+	}
+	return err
+}
+
+// getExistingSecret gets the K8s secret specified, and returns a bool indicating whether it exists, and the secret
+// itself. It returns nil if it doesn't exist.
+func (r *PeeringAcceptorController) getExistingSecret(ctx context.Context, name string, namespace string) (bool, *corev1.Secret, error) {
```
Comment (ishustava) [inline control-plane/connect-inject/peering_acceptor_controller.go:241]: Could we check if secret is nil to check if it exists instead of returning an additional boolean?

File: charts/consul/templates/crd-peeringacceptors.yaml [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+    release: {{ .Release.Name }}
+    component: crd
+spec:
+  group: consul.hashicorp.com
+  names:
+    kind: PeeringAcceptor
+    listKind: PeeringAcceptorList
+    plural: peeringacceptors
+    singular: peeringacceptor
+  scope: Namespaced
```
Comment (ishustava) [inline charts/consul/templates/crd-peeringacceptors.yaml:23]: should it be a cluster resource?

**Review by ishustava — COMMENTED** [review]
Looks good! Leaving my remaining comments. I think this is good to go once we decide what to do about some questions I left.

File: control-plane/api/v1alpha1/peeringacceptor_types.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (thisisnotashwin) [reply control-plane/api/v1alpha1/peeringacceptor_types.go]: sorry i didnt respond to this. i think using the object reference makes a lot of sense. it is a lot more detailed and specific about the secret that is being generated IMO.

File: control-plane/connect-inject/peering_acceptor_controller.go [hunk 17]
```diff
@@ -0,0 +1,335 @@
...
+	}
+
+	// TODO(peering): remove this when validation webhook exists.
+	if peeringAcceptor.Status.Secret.Backend != peeringAcceptor.Spec.Peer.Secret.Backend {
+		return false, false, errors.New("PeeringAcceptor backend cannot be changed")
+	}
+	// Compare the existing secret hash.
+	// Get the secret specified by the status, make sure it matches the status' secret.latestHash.
+	if existingSecret != nil {
+		existingSecretHashBytes := sha256.Sum256(existingSecret.Data[peeringAcceptor.Status.Secret.Key])
```
[the remark this answers is not in the document]
Comment (thisisnotashwin) [reply control-plane/connect-inject/peering_acceptor_controller.go:191]: I really like this idea!!

File: charts/consul/templates/crd-peeringacceptors.yaml (hunk 16 above)
Comment (thisisnotashwin) [reply charts/consul/templates/crd-peeringacceptors.yaml:23]: this was something that i had suggested. This to me felt similar to our config entries of proxy_defaults and mesh ie it is ok for these to be namespaced because there are users that prefer to save all of their config in a single configuration namespace and this feels like something that falls into the same category. additionally, given it is responsible for creating and managing a secret, it is easy to assume that the secret needs to be created in the same namespace as the resource itself instead of us having to determine which namespace the secret has to be written to.

Comment (ishustava) [reply charts/consul/templates/crd-peeringacceptors.yaml:23]: Ahh ok! I was thinking that you would only want one of these per cluster, and so that's why it might make sense from rbac perspective to make it cluster-wide. For secret, i think it's ok to manage a secret in some namespace (e.g. ClusterRoleBinding refers to a service account and service accounts are namespaced).

File: .github/workflows/test.yml (hunk 9 above)
Comment (ndhanushkodi) [reply .github/workflows/test.yml:142]: Yup agreed, we'd not want to cut a release with all this unless there's something for Consul also plus we'll want to address TODOs and unit tests (for peering dialer) on the feature branch `peering` before merging to main.

File: control-plane/connect-inject/peering_acceptor_controller.go (hunk 10 above)
Comment (ndhanushkodi) [reply control-plane/connect-inject/peering_acceptor_controller.go:45]: Yup! done!

File: charts/consul/templates/crd-peeringacceptors.yaml [hunk 18]
```diff
@@ -0,0 +1,99 @@
+{{- if .Values.connectInject.enabled }}
```
[the remark this answers is not in the document]
Comment (ndhanushkodi) [reply charts/consul/templates/crd-peeringacceptors.yaml:1]: Yeah that sounds good-- there is technically a service discovery case that can be enabled without service mesh, which could mean Consul DNS for imported services could work without enabling mesh functionality. Do you think it's worth enabling these controllers by default, not gated on any values?

File: charts/consul/templates/crd-peeringacceptors.yaml (hunk 16 above)
Comment (ndhanushkodi) [reply charts/consul/templates/crd-peeringacceptors.yaml:23]: Yeah I also think it might be weird for PeeringAcceptors to be cluster wide and for their managed secrets to be namespaced, it might be nicer to always create the secret in whatever ns the PeeringAcceptor is in. And yup there can be multiple PeeringAcceptors in a single cluster

File: control-plane/config/crd/patches/cainjection_in_peeringacceptors.yaml [hunk 19]
```diff
@@ -0,0 +1,8 @@
+# The following patch adds a directive for certmanager to inject CA into the CRD
+# CRD conversion requires k8s 1.13 or later.
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
```
[the remark this answers is not in the document]
Comment (ndhanushkodi) [reply control-plane/config/crd/patches/cainjection_in_peeringacceptors.yaml:4]: I'll remove since its not there for any other CRs

File: control-plane/config/crd/patches/webhook_in_peeringacceptors.yaml [hunk 20]
```diff
@@ -0,0 +1,17 @@
+# The following patch enables conversion webhook for CRD
+# CRD conversion requires k8s 1.13 or later.
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
```
[the remark this answers is not in the document]
Comment (ndhanushkodi) [reply control-plane/config/crd/patches/webhook_in_peeringacceptors.yaml:4]: I'll remove since its not there for any other CRs

File: control-plane/config/samples/consul_v1alpha1_peeringacceptor.yaml [hunk 21]
```diff
@@ -0,0 +1,11 @@
+apiVersion: consul.hashicorp.com/v1alpha1
+kind: PeeringAcceptor
```
[the remark this answers is not in the document]
Comment (ndhanushkodi) [reply control-plane/config/samples/consul_v1alpha1_peeringacceptor.yaml:2]: Yeah, it was a generated sample, I think we can remove

File: control-plane/config/rbac/peeringacceptor_editor_role.yaml [hunk 22]
```diff
@@ -0,0 +1,24 @@
+# permissions for end users to edit peeringacceptors.
```
Comment (ndhanushkodi) [inline control-plane/config/rbac/peeringacceptor_editor_role.yaml:1]: Everything in config/rbac was also generated but doesn't exist for other CRs, I think we can remove?

File: control-plane/connect-inject/peering_acceptor_controller.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (ndhanushkodi) [reply control-plane/connect-inject/peering_acceptor_controller.go:191]: Yup, great idea!

File: control-plane/connect-inject/peering_acceptor_controller.go (hunk 12 above)
Comment (ndhanushkodi) [reply control-plane/connect-inject/peering_acceptor_controller.go:214]: Omg yeah whoops I had done that so I could run tests using just the two file names (controller and controller_test) without needing to include the file that has pointerToBool! Thanks for catching!!

File: control-plane/api/v1alpha1/peeringacceptor_types.go [hunk 23]
```diff
@@ -0,0 +1,76 @@
...
+
+// PeeringAcceptorList contains a list of PeeringAcceptor.
+type PeeringAcceptorList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []PeeringAcceptor `json:"items"`
+}
+
+// PeeringAcceptorSpec defines the desired state of PeeringAcceptor.
+type PeeringAcceptorSpec struct {
```
[the remark this answers is not in the document]
Comment (ndhanushkodi) [reply control-plane/api/v1alpha1/peeringacceptor_types.go:35]: Yup, added!

File: control-plane/connect-inject/peering_acceptor_controller.go [hunk 24]
```diff
@@ -0,0 +1,335 @@
...
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// PeeringAcceptor resources determine whether to generate a new peering token in Consul and store it in the backend
+// specified in the spec. If the resource doesn't exist, the peering should be deleted in Consul. If the resource
+// exists, and a peering doesn't exist in Consul, it should be created. If the resource exists, and a peering does exist
+// in Consul, it should be reconciled. If the status of the resource does not match the current state of the specified
+// secret, generate a new token and store it according to the spec.
+func (r *PeeringAcceptorController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	_ = context.Background()
+	_ = r.Log.WithValues("peering", req.NamespacedName)
```
[the remark this answers is not in the document]
Comment (ndhanushkodi) [reply control-plane/connect-inject/peering_acceptor_controller.go:50]: Ah these are being passed in from the inject-connect command that starts the controllers, this was just leftover boilerplate, removed

File: control-plane/connect-inject/peering_acceptor_controller.go [hunk 25]
```diff
@@ -0,0 +1,335 @@
...
+	}
+
+	statusSecretSet := false
+	if peeringAcceptor.Status.Secret != nil {
+		statusSecretSet = true
+	}
+
+	// Look up existing secret.
+	var existingSecret *corev1.Secret
+	if statusSecretSet {
```
[the remark this answers is not in the document]
Comment (ndhanushkodi) [reply control-plane/connect-inject/peering_acceptor_controller.go:79]: Any time we create a secret, we do then set the status. However, if the secret was successfully created but the status didn't properly get updated, then the scenario you described could happen. However, that would return a reconcile error since `updateStatus` would return an error. Upon the next reconcile, `statusSecretSet` would be false, but the peering would exist in Consul, so it'll go into the case in L140 and try to generate a token again, so I think it'd be fine.

File: control-plane/connect-inject/peering_acceptor_controller.go [hunk 26]
```diff
@@ -0,0 +1,335 @@
...
+		_, existingSecret, err = r.getExistingSecret(ctx, peeringAcceptor.Status.Secret.Name, peeringAcceptor.Namespace)
+		if err != nil {
+			_ = r.updateStatusError(ctx, peeringAcceptor, err)
+			return ctrl.Result{}, err
```
[the remark this answers is not in the document]
Comment (ndhanushkodi) [reply control-plane/connect-inject/peering_acceptor_controller.go:83]: The getExistingSecret function does not return an error if the secret doesn't exist, it just sets the existingSecret return value to nil, so in this case, existingSecret will just be nil but that wouldn't be an error case. In L95 we check if existingSecret is nil, and the shouldGenerateToken function also checks if the existingSecret is nil before deciding what to do.

File: control-plane/connect-inject/peering_acceptor_controller.go [hunk 27]
```diff
@@ -0,0 +1,335 @@
...
+	// Compare the existing secret hash.
+	// Get the secret specified by the status, make sure it matches the status' secret.latestHash.
+	if existingSecret != nil {
+		existingSecretHashBytes := sha256.Sum256(existingSecret.Data[peeringAcceptor.Status.Secret.Key])
+		existingSecretHash := hex.EncodeToString(existingSecretHashBytes[:])
+		if existingSecretHash != peeringAcceptor.Status.Secret.LatestHash {
+			return true, false, nil
+		}
+
+	}
```
[the remark this answers is not in the document]
Comment (ndhanushkodi) [reply control-plane/connect-inject/peering_acceptor_controller.go:199]: This is such a great catch, yes thanks! If someone has deleted the secret specified in the status we could get to this case! Updated this.

File: control-plane/connect-inject/peering_acceptor_controller.go (hunk 15 above)
Comment (ndhanushkodi) [reply control-plane/connect-inject/peering_acceptor_controller.go:243]: Yup! Changed!

File: control-plane/connect-inject/peering_acceptor_controller.go (hunk 11 above)
Comment (ndhanushkodi) [reply control-plane/connect-inject/peering_acceptor_controller.go]: nice, done!

File: control-plane/connect-inject/peering_acceptor_controller.go (hunk 14 above)
Comment (ndhanushkodi) [reply control-plane/connect-inject/peering_acceptor_controller.go:225]: I don't think there's anything we can do to handle the error from this function. If we fail to update the status with the error message, we don't want the status update failure to be what's returned from the reconcile, instead we want the real error. I suppose we could have this function just not return an error, and document that we are ignoring any errors in this case? The only error case is if the K8s api is down when trying to make an update to the status, and instead the real error would be returned as a reconcile error, wdyt?

File: control-plane/connect-inject/peering_acceptor_controller.go (hunk 1 above)
Comment (ndhanushkodi) [reply control-plane/connect-inject/peering_acceptor_controller.go:206]: update: haven't gotten to this yet

File: control-plane/api/v1alpha1/peeringacceptor_types.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (ndhanushkodi) [reply control-plane/api/v1alpha1/peeringacceptor_types.go]: ahh, yeah, I should update these fields then. Not yet done, still working on it.

**Comment by ndhanushkodi** [comment]
Updates:

Hey @ishustava re-requesting review, while still addressing:
- secret's resourceVersion instead of computing a hash (need to fix tests for this case, and make sure it all works)

@thisisnotashwin I'm still addressing:
- using the object reference
- refactor to have getters on the CRD fields

File: control-plane/api/v1alpha1/peeringacceptor_types.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (ndhanushkodi) [reply control-plane/api/v1alpha1/peeringacceptor_types.go]: Hey @thisisnotashwin so when I was looking at the object reference type, it says to prefer to not use it and use local types instead: https://github.com/kubernetes/api/blob/be84346886a48ff00fc676f9cb37f10e79dfd0da/core/v1/types.go#L5624

File: control-plane/connect-inject/peering_acceptor_controller.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (ndhanushkodi) [reply control-plane/connect-inject/peering_acceptor_controller.go]: updated to use resource version! this was a great idea!

File: control-plane/api/v1alpha1/peeringacceptor_types.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (ndhanushkodi) [reply control-plane/api/v1alpha1/peeringacceptor_types.go]: I renamed the fields in the local type like the example in the godoc above, wdyt?

File: control-plane/connect-inject/peering_acceptor_controller.go (hunk 1 above)
Comment (ndhanushkodi) [reply control-plane/connect-inject/peering_acceptor_controller.go:206]: I ended up just adding an Asana to clean this up later in favor of merging sooner: https://app.asana.com/0/1201832349181086/1202342432709154

**Comment by ndhanushkodi** [comment]
Made a cleanup Asana to refactor to have getters on the CRD fields

Currently thinking to not using the object reference, and using the local type instead due to https://github.com/kubernetes/api/blob/be84346886a48ff00fc676f9cb37f10e79dfd0da/core/v1/types.go#L5624

File: control-plane/api/v1alpha1/peeringacceptor_types.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (thisisnotashwin) [reply control-plane/api/v1alpha1/peeringacceptor_types.go]: aha! i think that makes a lot of sense. thanks for doing this research Nitya!! this is excellent!

File: control-plane/connect-inject/peering_acceptor_controller.go (hunk 1 above)
Comment (thisisnotashwin) [reply control-plane/connect-inject/peering_acceptor_controller.go:206]: perfect! thanks Nitya!

File: charts/consul/templates/crd-peeringacceptors.yaml (hunk 18 above)
[the remark this answers is not in the document]
Comment (ishustava) [reply charts/consul/templates/crd-peeringacceptors.yaml:1]: yeah I think we will in the future, but probably not for this feature.
